docs/tech-spike-write-ups/get-an-identity-oidc-oauth-with-ruby.md

```diff
@@ -0,0 +1,126 @@
+
+While it was possible to eventually achieve the goals of the spike, a number of issues were identified while working with the gems:
+
+The documentation is poor
+Example: doorkeeper’s explanation page for what an “Access Grant” is is a blank document
+The libraries have hardcoded defaults that don’t allow for certain use cases
+Example: swd hardcodes HTTPS as the only protocol that it will allow for the discovery endpoint, which means it’s not trivial to develop this system locally, without resorting to monkey patching or forking
+The libraries are not maintained and actively looking for maintainers
+Example: doorkeeper-openid_connect [meta] Looking for maintainers! #89
+The libraries have counter-intuitive behaviours
+Example: the example configuration for omniauth_openid_connect does not include an `issuer` field, which leads to a cryptic SSL error. Debugging this required digging 3 layers deep into the dependencies to find the issue above, regarding the hardcoded HTTPS
+
+It’s expected that, for sufficiently advanced use cases, projects end up owning every part of the dependency chain. However, this simple spike required a significant investment digging into the library code, devising monkey patches, or just figuring out how things work in lieu of documentation. This is a strong signal that the tooling requires significant investment.
+
+There is a high risk that DFE-Digital will have to fork and maintain these gems. Alongside being a big time investment, making changes requires deep knowledge of OAuth and OIDC. Moreover, maintaining our own forks also means staying afloat of changes in the specification, as well as new security advisories, CVEs.
+Risk #3: Burden to build and ops overhead
+It was felt during the spike that there is no “shortcut” to building a quick version of a SSO provider, with good documentation, staging environments, bespoke integrations that might be slightly different for each client, a support team, oncall rota, and so forth.
+
+The closest prior art to a project that tackles the same scope in DFE Digital is DFE Signin. The “gut feeling” after the spike is that there is nothing inherent about our usecase that would make it “easier for us.”
+
+Reaching an MVP state requires a significant investment, with “not much to show” until it is fully done, as a half-baked SSO integration is one half away from being usable.
+Alternatives to building a Single Sign-on
+In light of the risks discussed above, some alternatives were thought of that would still solve the problems of code reuse / joined up user experience, while avoiding some of the thorny bits of OAuth and OIDC.
+
+# Approach #1: JWT token handover
+JSON Web Tokens (JWT) are a secure way of passing messages between two parties. A simple token exchange flow could be implemented both in Find a lost TRN and any client apps that would allow users to transfer data between the services in a secure way, such as their TRNs.
```

Comment:
```suggestion
### Approach #1: JWT token handover

JSON Web Tokens (JWT) are a secure way of passing messages between two parties. A simple token exchange flow could be implemented both in Find a lost TRN and any client apps that would allow users to transfer data between the services in a secure way, such as their TRNs.
```